crates/http/src/routes.rs

```diff
@@ -38,11 +38,29 @@ impl Router {
     /// If there are multiple possible components registered for the same route or
     /// wildcard, this returns the last entry in the component map.
     pub(crate) fn route(&self, p: &str) -> Result<&str> {
-        self.routes
-            .iter()
-            .rfind(|(rp, _)| rp.matches(p))
-            .map(|(_, c)| c.as_ref())
-            .with_context(|| format!("Cannot match route for path {}", p))
+        let matches = self.routes.iter().filter(|(rp, _)| rp.matches(p));
+
+        let mut best_match: (Option<&str>, usize) = (None, 0); // matched id and pattern length
+
+        for (rp, id) in matches {
+            match rp {
+                RoutePattern::Exact(_m) => {
+                    // Exact matching routes take precedence over wildcard matches.
+                    return Ok(id);
+                }
+                RoutePattern::Wildcard(m) => {
+                    // Check and find longest matching prefix of wildcard pattern.
+                    let (_id_opt, len) = best_match;
+                    if m.len() >= len {
+                        best_match = (Some(id), m.len());
```

Comment:
I glanced over this first time round, but just wanted to say I thought this was an insightful shortcut - I had been imagining having to parse slashes and count bits, but the "only look at things that match" strategy indeed means you only need to consider string length (maximum prefix).  Nice one!